File: apps/backend/services/sentry_service.py
"""Sentry error tracking service with HIPAA-compliant PHI scrubbing."""

import logging
import os
from typing import Any, Dict, Optional

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.logging import LoggingIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration

from utils.phi_scrubber import scrub_phi

logger = logging.getLogger(__name__)


class PHIScrubberProcessor:
    """Sentry processor to scrub PHI from error data before sending."""

    @staticmethod
    def scrub_event(
        event: Dict[str, Any], hint: Dict[str, Any]
    ) -> Optional[Dict[str, Any]]:
        """Scrub PHI from Sentry event data.

        Args:
            event: Sentry event data
            hint: Additional context from Sentry

        Returns:
            Scrubbed event data or None to drop the event
        """
        try:
            # Scrub exception messages
            if "exception" in event and "values" in event["exception"]:
                for exception in event["exception"]["values"]:
                    if "value" in exception:
                        exception["value"] = scrub_phi(exception["value"])

                    # Scrub stack trace frames
                    if (
                        "stacktrace" in exception
                        and "frames" in exception["stacktrace"]
                    ):
                        for frame in exception["stacktrace"]["frames"]:
                            if "vars" in frame:
                                frame["vars"] = scrub_phi(frame["vars"])
                            if "context_line" in frame:
                                frame["context_line"] = scrub_phi(frame["context_line"])
                            if "pre_context" in frame:
                                frame["pre_context"] = [
                                    scrub_phi(line) for line in frame["pre_context"]
                                ]
                            if "post_context" in frame:
                                frame["post_context"] = [
                                    scrub_phi(line) for line in frame["post_context"]
                                ]

            # Scrub request data
            if "request" in event:
                request_data = event["request"]
                if "data" in request_data:
                    request_data["data"] = scrub_phi(request_data["data"])
                if "query_string" in request_data:
                    request_data["query_string"] = scrub_phi(
                        request_data["query_string"]
                    )
                if "headers" in request_data:
                    # Remove sensitive headers
                    sensitive_headers = ["authorization", "cookie", "x-api-key"]
                    for header in sensitive_headers:
                        request_data["headers"].pop(header, None)

            # Scrub user context (anonymize while keeping session tracking)
            if "user" in event:
                user_data = event["user"]
                if "id" in user_data:
                    # Keep user ID for session tracking but anonymize
                    user_data["id"] = f"user_{hash(user_data['id']) % 10000:04d}"
                # Remove other potentially identifying user data
                user_data.pop("email", None)
                user_data.pop("username", None)
                user_data.pop("name", None)

            # Scrub extra context data
            if "extra" in event:
                event["extra"] = scrub_phi(event["extra"])

            # Scrub breadcrumbs
            if "breadcrumbs" in event and "values" in event["breadcrumbs"]:
                for breadcrumb in event["breadcrumbs"]["values"]:
                    if "message" in breadcrumb:
                        breadcrumb["message"] = scrub_phi(breadcrumb["message"])
                    if "data" in breadcrumb:
                        breadcrumb["data"] = scrub_phi(breadcrumb["data"])

            return event

        except Exception as e:
            logger.error(f"Error scrubbing PHI from Sentry event: {e}")
            # Return None to drop the event if scrubbing fails
            return None


class SentryService:
    """Service for configuring and managing Sentry error tracking."""

    def __init__(self):
        self.dsn = os.getenv("SENTRY_DSN")
        self.environment = os.getenv("SENTRY_ENVIRONMENT", "development")
        self.release = os.getenv("SENTRY_RELEASE", "unknown")
        self.traces_sample_rate = float(os.getenv("SENTRY_TRACES_SAMPLE_RATE", "0.1"))
        self.profiles_sample_rate = float(
            os.getenv("SENTRY_PROFILES_SAMPLE_RATE", "0.1")
        )
        self.initialized = False

    def initialize(self) -> bool:
        """Initialize Sentry with HIPAA-compliant configuration.

        Returns:
            True if initialization successful, False otherwise
        """
        if not self.dsn:
            logger.warning("Sentry DSN not configured, error tracking disabled")
            return False

        try:
            sentry_sdk.init(
                dsn=self.dsn,
                environment=self.environment,
                release=self.release,
                traces_sample_rate=self.traces_sample_rate,
                profiles_sample_rate=self.profiles_sample_rate,
                integrations=[
                    FastApiIntegration(),
                    LoggingIntegration(level=logging.INFO, event_level=logging.ERROR),
                    SqlalchemyIntegration(),
                ],
                before_send=PHIScrubberProcessor.scrub_event,
                # Additional HIPAA compliance settings
                send_default_pii=False,  # Never send PII
                attach_stacktrace=True,
                max_breadcrumbs=50,
                debug=self.environment == "development",
            )

            # Set global tags
            sentry_sdk.set_tag("service", "pms-backend")
            sentry_sdk.set_tag("compliance", "hipaa")

            self.initialized = True
            logger.info(f"Sentry initialized for environment: {self.environment}")
            return True

        except Exception as e:
            logger.error(f"Failed to initialize Sentry: {e}")
            return False

    def capture_exception(
        self,
        exception: Exception,
        user_id: Optional[str] = None,
        correlation_id: Optional[str] = None,
        extra_context: Optional[Dict[str, Any]] = None,
    ) -> Optional[str]:
        """Capture an exception with additional context.

        Args:
            exception: The exception to capture
            user_id: Anonymous user identifier
            correlation_id: Request correlation ID
            extra_context: Additional context (will be PHI-scrubbed)

        Returns:
            Sentry event ID if captured, None otherwise
        """
        if not self.initialized:
            logger.error("Sentry not initialized, cannot capture exception")
            return None

        try:
            with sentry_sdk.push_scope() as scope:
                # Set user context (anonymized)
                if user_id:
                    scope.set_user({"id": f"user_{hash(user_id) % 10000:04d}"})

                # Set correlation ID for request tracking
                if correlation_id:
                    scope.set_tag("correlation_id", correlation_id)

                # Add extra context (will be scrubbed by processor)
                if extra_context:
                    for key, value in extra_context.items():
                        scope.set_extra(key, value)

                # Capture the exception
                event_id = sentry_sdk.capture_exception(exception)
                logger.info(f"Exception captured by Sentry: {event_id}")
                return event_id

        except Exception as e:
            logger.error(f"Failed to capture exception in Sentry: {e}")
            return None

    def capture_message(
        self,
        message: str,
        level: str = "info",
        user_id: Optional[str] = None,
        correlation_id: Optional[str] = None,
        extra_context: Optional[Dict[str, Any]] = None,
    ) -> Optional[str]:
        """Capture a message with additional context.

        Args:
            message: The message to capture (will be PHI-scrubbed)
            level: Message level (debug, info, warning, error, fatal)
            user_id: Anonymous user identifier
            correlation_id: Request correlation ID
            extra_context: Additional context (will be PHI-scrubbed)

        Returns:
            Sentry event ID if captured, None otherwise
        """
        if not self.initialized:
            return None

        try:
            with sentry_sdk.push_scope() as scope:
                # Set user context (anonymized)
                if user_id:
                    scope.set_user({"id": f"user_{hash(user_id) % 10000:04d}"})

                # Set correlation ID
                if correlation_id:
                    scope.set_tag("correlation_id", correlation_id)

                # Add extra context
                if extra_context:
                    for key, value in extra_context.items():
                        scope.set_extra(key, value)

                # Capture the message (will be scrubbed by processor)
                event_id = sentry_sdk.capture_message(scrub_phi(message), level)
                return event_id

        except Exception as e:
            logger.error(f"Failed to capture message in Sentry: {e}")
            return None

    def create_test_event(self) -> Optional[str]:
        """Create a test event for validation.

        Returns:
            Sentry event ID if successful, None otherwise
        """
        try:
            test_exception = Exception(
                "Test exception for Sentry validation - PHI scrubbing test: "
                "John Doe SSN 123-45-6789"
            )
            return self.capture_exception(
                test_exception,
                user_id="test_user_123",
                correlation_id="test_correlation_id",
                extra_context={
                    "test_data": "Patient: Jane Smith, DOB: 1990-01-01",
                    "environment": self.environment,
                    "test_type": "validation",
                },
            )
        except Exception as e:
            logger.error(f"Failed to create test event: {e}")
            return None


# Global Sentry service instance
sentry_service = SentryService()


def get_sentry_service() -> SentryService:
    """Get the global Sentry service instance."""
    return sentry_service

File: apps/frontend/src/services/sentryService.ts
/**
 * Sentry service for client-side error tracking with PHI scrubbing
 */

import * as Sentry from "@sentry/react";
import { BrowserTracing } from "@sentry/tracing";
import { scrubPHI } from "../utils/phiScrubber";

// PHI patterns for client-side scrubbing
const PHI_PATTERNS = [
  // SSN patterns
  /\b\d{3}-\d{2}-\d{4}\b/g,
  /\b\d{9}\b/g,
  // Phone numbers
  /\b\d{3}-\d{3}-\d{4}\b/g,
  /\(\d{3}\)\s*\d{3}-\d{4}/g,
  // Email addresses
  /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g,
  // Date of birth patterns
  /\b\d{1,2}\/\d{1,2}\/\d{4}\b/g,
  /\b\d{4}-\d{2}-\d{2}\b/g,
  // Medical record numbers
  /\bMRN[:\s]*\d+/gi,
  /\bMedical Record[:\s]*\d+/gi,
  // Patient names (common patterns)
  /\bPatient[:\s]+[A-Za-z]+\s+[A-Za-z]+/gi,
  /\bDOB[:\s]+\d{1,2}\/\d{1,2}\/\d{4}/gi,
];

/**
 * Scrub PHI from text content
 */
function scrubPHIFromText(text: string): string {
  let scrubbed = text;

  PHI_PATTERNS.forEach((pattern) => {
    scrubbed = scrubbed.replace(pattern, "[REDACTED]");
  });

  return scrubbed;
}

/**
 * Scrub PHI from Sentry event data
 */
function scrubSentryEvent(event: Sentry.Event): Sentry.Event {
  // Scrub exception messages
  if (event.exception?.values) {
    event.exception.values.forEach((exception) => {
      if (exception.value) {
        exception.value = scrubPHIFromText(exception.value);
      }

      // Scrub stack trace
      if (exception.stacktrace?.frames) {
        exception.stacktrace.frames.forEach((frame) => {
          if (frame.filename) {
            frame.filename = scrubPHIFromText(frame.filename);
          }
          if (frame.function) {
            frame.function = scrubPHIFromText(frame.function);
          }
          if (frame.vars) {
            Object.keys(frame.vars).forEach((key) => {
              if (typeof frame.vars![key] === "string") {
                frame.vars![key] = scrubPHIFromText(frame.vars![key]);
              }
            });
          }
        });
      }
    });
  }

  // Scrub message
  if (event.message) {
    event.message = scrubPHIFromText(event.message);
  }

  // Scrub breadcrumbs
  if (event.breadcrumbs) {
    event.breadcrumbs.forEach((breadcrumb) => {
      if (breadcrumb.message) {
        breadcrumb.message = scrubPHIFromText(breadcrumb.message);
      }
      if (breadcrumb.data) {
        Object.keys(breadcrumb.data).forEach((key) => {
          if (typeof breadcrumb.data![key] === "string") {
            breadcrumb.data![key] = scrubPHIFromText(breadcrumb.data![key]);
          }
        });
      }
    });
  }

  // Scrub request data
  if (event.request) {
    if (event.request.url) {
      event.request.url = scrubPHIFromText(event.request.url);
    }
    if (event.request.query_string) {
      event.request.query_string = scrubPHIFromText(event.request.query_string);
    }
    if (event.request.data) {
      if (typeof event.request.data === "string") {
        event.request.data = scrubPHIFromText(event.request.data);
      }
    }

    // Remove sensitive headers
    if (event.request.headers) {
      const sensitiveHeaders = ["authorization", "cookie", "x-api-key"];
      sensitiveHeaders.forEach((header) => {
        if (event.request!.headers![header]) {
          event.request!.headers![header] = "[REDACTED]";
        }
      });
    }
  }

  // Scrub user data (anonymize)
  if (event.user) {
    if (event.user.id) {
      // Hash user ID for anonymization
      const userHash = Math.abs(hashCode(event.user.id.toString())) % 10000;
      event.user.id = `user_${userHash.toString().padStart(4, "0")}`;
    }

    // Remove sensitive user fields
    delete event.user.email;
    delete event.user.username;
    delete event.user.ip_address;
  }

  // Scrub extra context
  if (event.extra) {
    Object.keys(event.extra).forEach((key) => {
      if (typeof event.extra![key] === "string") {
        event.extra![key] = scrubPHIFromText(event.extra![key]);
      }
    });
  }

  return event;
}

/**
 * Simple hash function for anonymization
 */
function hashCode(str: string): number {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash = hash & hash; // Convert to 32-bit integer
  }
  return hash;
}

/**
 * Initialize Sentry for the React application
 */
export function initSentry(): void {
  const dsn = process.env.REACT_APP_SENTRY_DSN;
  const environment = process.env.REACT_APP_SENTRY_ENVIRONMENT || "development";
  const release = process.env.REACT_APP_SENTRY_RELEASE;

  if (!dsn) {
    console.warn("Sentry DSN not configured, error tracking disabled");
    return;
  }

  Sentry.init({
    dsn,
    environment,
    release,
    integrations: [
      new BrowserTracing({
        // Trace navigation and user interactions
        routingInstrumentation: Sentry.reactRouterV6Instrumentation(
          React.useEffect,
          useLocation,
          useNavigationType,
          createRoutesFromChildren,
          matchRoutes,
        ),
      }),
    ],

    // Performance monitoring
    tracesSampleRate: parseFloat(process.env.REACT_APP_SENTRY_TRACES_SAMPLE_RATE || "0.1"),

    // Session replay (disabled for HIPAA compliance)
    replaysSessionSampleRate: 0.0,
    replaysOnErrorSampleRate: 0.0,

    // PHI scrubbing before send
    beforeSend(event) {
      return scrubSentryEvent(event);
    },

    // Additional configuration
    debug: environment === "development",

    // Ignore common non-critical errors
    ignoreErrors: [
      "ResizeObserver loop limit exceeded",
      "Non-Error promise rejection captured",
      "ChunkLoadError",
      "Loading chunk",
      "Network Error",
    ],

    // Set initial user context (anonymized)
    initialScope: {
      tags: {
        component: "pms-frontend",
      },
    },
  });

  console.log(`Sentry initialized for environment: ${environment}`);
}

/**
 * Set user context with anonymization
 */
export function setSentryUser(userId: string, additionalData?: Record<string, any>): void {
  Sentry.setUser({
    id: `user_${Math.abs(hashCode(userId)) % 10000}`,
    ...additionalData,
  });
}

/**
 * Add breadcrumb with PHI scrubbing
 */
export function addSentryBreadcrumb(
  message: string,
  category: string = "custom",
  level: Sentry.SeverityLevel = "info",
  data?: Record<string, any>,
): void {
  Sentry.addBreadcrumb({
    message: scrubPHIFromText(message),
    category,
    level,
    data: data ? scrubPHIFromText(JSON.stringify(data)) : undefined,
  });
}

/**
 * Capture exception with PHI scrubbing
 */
export function captureSentryException(error: Error, context?: Record<string, any>): void {
  Sentry.withScope((scope) => {
    if (context) {
      Object.keys(context).forEach((key) => {
        const value = context[key];
        const scrubbed = typeof value === "string" ? scrubPHIFromText(value) : value;
        scope.setContext(key, scrubbed);
      });
    }

    Sentry.captureException(error);
  });
}

/**
 * Capture message with PHI scrubbing
 */
export function captureSentryMessage(
  message: string,
  level: Sentry.SeverityLevel = "info",
  context?: Record<string, any>,
): void {
  Sentry.withScope((scope) => {
    if (context) {
      Object.keys(context).forEach((key) => {
        const value = context[key];
        const scrubbed = typeof value === "string" ? scrubPHIFromText(value) : value;
        scope.setContext(key, scrubbed);
      });
    }

    Sentry.captureMessage(scrubPHIFromText(message), level);
  });
}

/**
 * Create test error for validation
 */
export function createSentryTestError(): void {
  const testError = new Error(
    "Test error for Sentry validation. PHI test: Patient John Doe SSN 123-45-6789",
  );

  captureSentryException(testError, {
    test_type: "sentry_validation",
    timestamp: Date.now(),
    test_data: "Patient: Jane Smith, DOB: 1990-01-01",
  });
}

// Import React Router hooks (these would be imported from react-router-dom)
// This is a placeholder - actual imports would be:
// import { useEffect } from 'react';
// import { useLocation, useNavigationType, createRoutesFromChildren, matchRoutes } from 'react-router-dom';
const React = { useEffect: () => {} };
const useLocation = () => ({});
const useNavigationType = () => "POP";
const createRoutesFromChildren = () => [];
const matchRoutes = () => [];
